chain-core/src/mls/mod.rs

```diff
@@ -1,32 +1,82 @@
-use crate::tx::TransactionId;
+use crate::tx::{data::TxId, TransactionId, TransactionType};
 use parity_scale_codec::{Decode, Encode};
 use std::prelude::v1::Vec;
 
+/// message to remove members
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct CommitRemoveTx {
+    proposals: Vec<Vec<u8>>, // [MLSPlaintext] -- Remove
+    commit: Vec<u8>,         // MLSPlaintext -- Commit
+}
+
+impl TransactionId for CommitRemoveTx {}
+
+impl TransactionType for CommitRemoveTx {
+    /// NOTE: this tag is used in tx id calculations / consensus critical,
+    /// so even if type is renamed or otherwise refactored,
+    /// this tag MUST remain the same
+    fn tx_type_tag() -> &'static [u8] {
+        b"commitremovetx"
```

Comment:
```rust
enum TaggedTx {
  WithdrawTx(WithdrawTx),
  DepositTx(DepositTx),
  TransferTx(TransferTx),
  ... repeat for all possible tx types
}

impl TaggedTx {
  fn id() {
    // the tx tags can be kept together.
  }
}

impl From<WithdrawTx> for TaggedTx {
  ;
}

impl<T: Into<TaggedTx>> TransactionId for T {
  fn id(&self) {
    TaggedTx::from(self).id()
  }
}
```

How about sth like this?

Reply:
yes, that's also possible -- it'd make it more feasible to use a single byte tag (I avoided that here, as there's no single ordering on tx types)

Reply:
tried it -- just a sketch -- two small annoyances with this one are 1) clone required (I can try to rewrite it with `TaggedTx<'a> { Transfer(&'a ...` if that would work), 2) to keep the old behaviour under a feature-flag, there'll need to be feature guards on all transactions (to import and `impl TransactionId`)

```
diff --git a/chain-core/src/mls/mod.rs b/chain-core/src/mls/mod.rs
index 02230b3c..e3d05550 100644
--- a/chain-core/src/mls/mod.rs
+++ b/chain-core/src/mls/mod.rs
@@ -1,4 +1,4 @@
-use crate::tx::{data::TxId, TransactionId, TransactionType};
+use crate::tx::{data::TxId, TaggedTransaction, TransactionId};
 use parity_scale_codec::{Decode, Encode};
 use std::prelude::v1::Vec;
 
@@ -9,14 +9,9 @@ pub struct CommitRemoveTx {
     commit: Vec<u8>,         // MLSPlaintext -- Commit
 }
 
-impl TransactionId for CommitRemoveTx {}
-
-impl TransactionType for CommitRemoveTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"commitremovetx"
+impl From<CommitRemoveTx> for TaggedTransaction {
+    fn from(tx: CommitRemoveTx) -> TaggedTransaction {
+        TaggedTransaction::MLSRemoveCommitProposal(tx)
     }
 }
 
@@ -27,14 +22,9 @@ pub struct SelfUpdateProposalTx {
     commit: Vec<u8>,   // MLSPlaintext -- Commit
 }
 
-impl TransactionId for SelfUpdateProposalTx {}
-
-impl TransactionType for SelfUpdateProposalTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same    
-    fn tx_type_tag() -> &'static [u8] {
-        b"selfupdateproposaltx"
+impl From<SelfUpdateProposalTx> for TaggedTransaction {
+    fn from(tx: SelfUpdateProposalTx) -> TaggedTransaction {
+        TaggedTransaction::MLSSelfUpdateProposal(tx)
     }
 }
 
@@ -42,14 +32,9 @@ impl TransactionType for SelfUpdateProposalTx {
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
 pub struct NackMsgTx(Vec<u8>);
 
-impl TransactionId for NackMsgTx {}
-
-impl TransactionType for NackMsgTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same    
-    fn tx_type_tag() -> &'static [u8] {
-        b"nackmsgtx"
+impl From<NackMsgTx> for TaggedTransaction {
+    fn from(tx: NackMsgTx) -> TaggedTransaction {
+        TaggedTransaction::MLSMsgNack(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/deposit.rs b/chain-core/src/state/account/op/data/deposit.rs
index 15030393..eae4a6cd 100644
--- a/chain-core/src/state/account/op/data/deposit.rs
+++ b/chain-core/src/state/account/op/data/deposit.rs
@@ -1,7 +1,7 @@
 use crate::state::account::address::StakedStateAddress;
 use crate::state::account::op::data::attribute::StakedStateOpAttributes;
 use crate::tx::data::input::TxoPointer;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 use serde::{Deserialize, Serialize};
 use std::fmt;
@@ -59,14 +59,9 @@ impl Encode for DepositBondTx {
     }
 }
 
-impl TransactionId for DepositBondTx {}
-
-impl TransactionType for DepositBondTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"deposittx"
+impl From<DepositBondTx> for TaggedTransaction {
+    fn from(tx: DepositBondTx) -> TaggedTransaction {
+        TaggedTransaction::Deposit(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/unbond.rs b/chain-core/src/state/account/op/data/unbond.rs
index 14606dd0..2043af6b 100644
--- a/chain-core/src/state/account/op/data/unbond.rs
+++ b/chain-core/src/state/account/op/data/unbond.rs
@@ -2,7 +2,7 @@ use crate::init::coin::Coin;
 use crate::state::account::address::StakedStateAddress;
 use crate::state::account::op::data::attribute::StakedStateOpAttributes;
 use crate::state::account::Nonce;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -55,14 +55,9 @@ impl Encode for UnbondTx {
     }
 }
 
-impl TransactionId for UnbondTx {}
-
-impl TransactionType for UnbondTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"unbondtx"
+impl From<UnbondTx> for TaggedTransaction {
+    fn from(tx: UnbondTx) -> TaggedTransaction {
+        TaggedTransaction::UnbondStakeTx(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/withdraw.rs b/chain-core/src/state/account/op/data/withdraw.rs
index d3e0c3d8..43bb84a4 100644
--- a/chain-core/src/state/account/op/data/withdraw.rs
+++ b/chain-core/src/state/account/op/data/withdraw.rs
@@ -2,7 +2,7 @@ use crate::init::coin::{sum_coins, Coin, CoinError};
 use crate::state::account::Nonce;
 use crate::tx::data::attribute::TxAttributes;
 use crate::tx::data::output::TxOut;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -48,14 +48,9 @@ impl Encode for WithdrawUnbondedTx {
     }
 }
 
-impl TransactionId for WithdrawUnbondedTx {}
-
-impl TransactionType for WithdrawUnbondedTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"withdrawtx"
+impl From<WithdrawUnbondedTx> for TaggedTransaction {
+    fn from(tx: WithdrawUnbondedTx) -> TaggedTransaction {
+        TaggedTransaction::Withdraw(tx)
     }
 }
 
diff --git a/chain-core/src/state/validator/nodejoin.rs b/chain-core/src/state/validator/nodejoin.rs
index 3f690dfd..514f144c 100644
--- a/chain-core/src/state/validator/nodejoin.rs
+++ b/chain-core/src/state/validator/nodejoin.rs
@@ -1,5 +1,5 @@
 use crate::state::account::{NodeMetadata, Nonce, StakedStateAddress, StakedStateOpAttributes};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -60,14 +60,9 @@ impl Encode for NodeJoinRequestTx {
     }
 }
 
-impl TransactionId for NodeJoinRequestTx {}
-
-impl TransactionType for NodeJoinRequestTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"nodejointx"
+impl From<NodeJoinRequestTx> for TaggedTransaction {
+    fn from(tx: NodeJoinRequestTx) -> TaggedTransaction {
+        TaggedTransaction::NodeJoinTx(tx)
     }
 }
 
diff --git a/chain-core/src/state/validator/unjail.rs b/chain-core/src/state/validator/unjail.rs
index 6e38d00c..f3b17d47 100644
--- a/chain-core/src/state/validator/unjail.rs
+++ b/chain-core/src/state/validator/unjail.rs
@@ -1,5 +1,5 @@
 use crate::state::account::{Nonce, StakedStateAddress, StakedStateOpAttributes};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -43,14 +43,9 @@ impl Encode for UnjailTx {
     }
 }
 
-impl TransactionId for UnjailTx {}
-
-impl TransactionType for UnjailTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"unjailtx"
+impl From<UnjailTx> for TaggedTransaction {
+    fn from(tx: UnjailTx) -> TaggedTransaction {
+        TaggedTransaction::UnjailTx(tx)
     }
 }
 
diff --git a/chain-core/src/tx/data/mod.rs b/chain-core/src/tx/data/mod.rs
index 3f751a2a..adbcc633 100644
--- a/chain-core/src/tx/data/mod.rs
+++ b/chain-core/src/tx/data/mod.rs
@@ -20,7 +20,7 @@ use serde::{Deserialize, Serialize};
 use crate::common::H256;
 use crate::init::coin::{sum_coins, Coin, CoinError};
 use crate::tx::data::{attribute::TxAttributes, input::TxoPointer, output::TxOut};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 
 /// Each input is 34 bytes
 /// Each output is 33 (address) + 8 (amount) + 9 (timelock) = 50 bytes
@@ -30,15 +30,6 @@ use crate::tx::{TransactionId, TransactionType};
 /// So, maximum transaction size (34 * 64) + (50 * 64) + 2688 = 8064
 const MAX_TX_SIZE: usize = 8100; // 8100 bytes
 
-/// Calculates hash of the input data -- if SCALE-serialized TX is passed in, it's equivalent to TxId.
-/// It uses blake3.
-pub fn txid_hash<T: TransactionId>(buf: &T) -> H256 {
-    let mut hasher = blake3::Hasher::new();
-    hasher.update(T::tx_type_tag());
-    hasher.update(&buf.encode());
-    hasher.finalize().into()
-}
-
 /// Key to identify the used TXID hash function, e.g. in ProofOps.
 pub const TXID_HASH_ID: &[u8; 6] = b"blake3";
 
@@ -102,14 +93,9 @@ impl fmt::Display for Tx {
     }
 }
 
-impl TransactionId for Tx {}
-
-impl TransactionType for Tx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"transfertx"
+impl From<Tx> for TaggedTransaction {
+    fn from(tx: Tx) -> TaggedTransaction {
+        TaggedTransaction::Transfer(tx)
     }
 }
 
diff --git a/chain-core/src/tx/mod.rs b/chain-core/src/tx/mod.rs
index 86104934..47d2a0c8 100644
--- a/chain-core/src/tx/mod.rs
+++ b/chain-core/src/tx/mod.rs
@@ -20,8 +20,6 @@ use crate::state::account::{
 };
 use crate::state::tendermint::BlockHeight;
 use crate::state::validator::NodeJoinRequestTx;
-#[cfg(not(feature = "legacy-txid"))]
-use crate::tx::data::txid_hash;
 use crate::tx::data::TxId;
 use aead::Payload;
 use data::input::{TxoPointer, TxoSize};
@@ -546,29 +544,56 @@ impl Decode for TxAux {
 /// or make sure you override the default implementation to take the identifier
 /// from the transaction data type.
 #[cfg(not(feature = "legacy-txid"))]
-pub trait TransactionId: Encode + Sized + TransactionType {
+pub trait TransactionId {
     /// retrieves a TX ID (currently blake3(<tx type tag> || scale_codec_bytes(tx)))
+    fn id(&self) -> TxId;
+}
+
+impl<T: Into<TaggedTransaction> + Clone> TransactionId for T {
     fn id(&self) -> TxId {
-        txid_hash(self)
+        let tx: TaggedTransaction = self.clone().into();
+        TaggedTransaction::from(tx).id()
     }
 }
 
 /// 0.5-compatible version: This trait is for transaction data type to define the "txid"
 /// that's used as a message digest in signing.
 #[cfg(feature = "legacy-txid")]
-pub trait TransactionId: Encode + Sized + TransactionType {
+pub trait TransactionId: Encode {
     /// 0.5-compatible version: retrieves a TX ID (currently blake3(scale_codec_bytes(tx)))
     fn id(&self) -> TxId {
         blake3::hash(&self.encode()).into()
     }
 }
 
-/// type tag used in transaction identifier calculations
-pub trait TransactionType {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8];
+/// used for TXID calculation -- contains all possible tx types
+/// NOTE: do not reorder, as the byte tag is used in txid calculation
+#[derive(Encode)]
+pub enum TaggedTransaction {
+    /// transfer transaction
+    Transfer(Tx),
+    /// deposit stake to bonded amount
+    Deposit(DepositBondTx),
+    /// withdraw unbonded amount
+    Withdraw(WithdrawUnbondedTx),
+    /// unbond stake
+    UnbondStakeTx(UnbondTx),
+    /// unjail request
+    UnjailTx(UnjailTx),
+    /// node join request
+    NodeJoinTx(NodeJoinRequestTx),
+    /// removal proposals + commit
+    MLSRemoveCommitProposal(crate::mls::CommitRemoveTx),
+    /// update proposal + commit
+    MLSSelfUpdateProposal(crate::mls::SelfUpdateProposalTx),
+    /// NACK
+    MLSMsgNack(crate::mls::NackMsgTx),
+}
+
+impl TaggedTransaction {
+    fn id(&self) -> TxId {
+        blake3::hash(&self.encode()).into()
+    }
 }
 
 impl TxAux {
tomas@MacBook-Pro chain-core % git diff > 
tomas@MacBook-Pro chain-core % cat test-single.patch 
diff --git a/chain-core/src/mls/mod.rs b/chain-core/src/mls/mod.rs
index 02230b3c..e3d05550 100644
--- a/chain-core/src/mls/mod.rs
+++ b/chain-core/src/mls/mod.rs
@@ -1,4 +1,4 @@
-use crate::tx::{data::TxId, TransactionId, TransactionType};
+use crate::tx::{data::TxId, TaggedTransaction, TransactionId};
 use parity_scale_codec::{Decode, Encode};
 use std::prelude::v1::Vec;
 
@@ -9,14 +9,9 @@ pub struct CommitRemoveTx {
     commit: Vec<u8>,         // MLSPlaintext -- Commit
 }
 
-impl TransactionId for CommitRemoveTx {}
-
-impl TransactionType for CommitRemoveTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"commitremovetx"
+impl From<CommitRemoveTx> for TaggedTransaction {
+    fn from(tx: CommitRemoveTx) -> TaggedTransaction {
+        TaggedTransaction::MLSRemoveCommitProposal(tx)
     }
 }
 
@@ -27,14 +22,9 @@ pub struct SelfUpdateProposalTx {
     commit: Vec<u8>,   // MLSPlaintext -- Commit
 }
 
-impl TransactionId for SelfUpdateProposalTx {}
-
-impl TransactionType for SelfUpdateProposalTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same    
-    fn tx_type_tag() -> &'static [u8] {
-        b"selfupdateproposaltx"
+impl From<SelfUpdateProposalTx> for TaggedTransaction {
+    fn from(tx: SelfUpdateProposalTx) -> TaggedTransaction {
+        TaggedTransaction::MLSSelfUpdateProposal(tx)
     }
 }
 
@@ -42,14 +32,9 @@ impl TransactionType for SelfUpdateProposalTx {
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
 pub struct NackMsgTx(Vec<u8>);
 
-impl TransactionId for NackMsgTx {}
-
-impl TransactionType for NackMsgTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same    
-    fn tx_type_tag() -> &'static [u8] {
-        b"nackmsgtx"
+impl From<NackMsgTx> for TaggedTransaction {
+    fn from(tx: NackMsgTx) -> TaggedTransaction {
+        TaggedTransaction::MLSMsgNack(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/deposit.rs b/chain-core/src/state/account/op/data/deposit.rs
index 15030393..eae4a6cd 100644
--- a/chain-core/src/state/account/op/data/deposit.rs
+++ b/chain-core/src/state/account/op/data/deposit.rs
@@ -1,7 +1,7 @@
 use crate::state::account::address::StakedStateAddress;
 use crate::state::account::op::data::attribute::StakedStateOpAttributes;
 use crate::tx::data::input::TxoPointer;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 use serde::{Deserialize, Serialize};
 use std::fmt;
@@ -59,14 +59,9 @@ impl Encode for DepositBondTx {
     }
 }
 
-impl TransactionId for DepositBondTx {}
-
-impl TransactionType for DepositBondTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"deposittx"
+impl From<DepositBondTx> for TaggedTransaction {
+    fn from(tx: DepositBondTx) -> TaggedTransaction {
+        TaggedTransaction::Deposit(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/unbond.rs b/chain-core/src/state/account/op/data/unbond.rs
index 14606dd0..2043af6b 100644
--- a/chain-core/src/state/account/op/data/unbond.rs
+++ b/chain-core/src/state/account/op/data/unbond.rs
@@ -2,7 +2,7 @@ use crate::init::coin::Coin;
 use crate::state::account::address::StakedStateAddress;
 use crate::state::account::op::data::attribute::StakedStateOpAttributes;
 use crate::state::account::Nonce;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -55,14 +55,9 @@ impl Encode for UnbondTx {
     }
 }
 
-impl TransactionId for UnbondTx {}
-
-impl TransactionType for UnbondTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"unbondtx"
+impl From<UnbondTx> for TaggedTransaction {
+    fn from(tx: UnbondTx) -> TaggedTransaction {
+        TaggedTransaction::UnbondStakeTx(tx)
     }
 }
 
diff --git a/chain-core/src/state/account/op/data/withdraw.rs b/chain-core/src/state/account/op/data/withdraw.rs
index d3e0c3d8..43bb84a4 100644
--- a/chain-core/src/state/account/op/data/withdraw.rs
+++ b/chain-core/src/state/account/op/data/withdraw.rs
@@ -2,7 +2,7 @@ use crate::init::coin::{sum_coins, Coin, CoinError};
 use crate::state::account::Nonce;
 use crate::tx::data::attribute::TxAttributes;
 use crate::tx::data::output::TxOut;
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -48,14 +48,9 @@ impl Encode for WithdrawUnbondedTx {
     }
 }
 
-impl TransactionId for WithdrawUnbondedTx {}
-
-impl TransactionType for WithdrawUnbondedTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"withdrawtx"
+impl From<WithdrawUnbondedTx> for TaggedTransaction {
+    fn from(tx: WithdrawUnbondedTx) -> TaggedTransaction {
+        TaggedTransaction::Withdraw(tx)
     }
 }
 
diff --git a/chain-core/src/state/validator/nodejoin.rs b/chain-core/src/state/validator/nodejoin.rs
index 3f690dfd..514f144c 100644
--- a/chain-core/src/state/validator/nodejoin.rs
+++ b/chain-core/src/state/validator/nodejoin.rs
@@ -1,5 +1,5 @@
 use crate::state::account::{NodeMetadata, Nonce, StakedStateAddress, StakedStateOpAttributes};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -60,14 +60,9 @@ impl Encode for NodeJoinRequestTx {
     }
 }
 
-impl TransactionId for NodeJoinRequestTx {}
-
-impl TransactionType for NodeJoinRequestTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"nodejointx"
+impl From<NodeJoinRequestTx> for TaggedTransaction {
+    fn from(tx: NodeJoinRequestTx) -> TaggedTransaction {
+        TaggedTransaction::NodeJoinTx(tx)
     }
 }
 
diff --git a/chain-core/src/state/validator/unjail.rs b/chain-core/src/state/validator/unjail.rs
index 6e38d00c..f3b17d47 100644
--- a/chain-core/src/state/validator/unjail.rs
+++ b/chain-core/src/state/validator/unjail.rs
@@ -1,5 +1,5 @@
 use crate::state::account::{Nonce, StakedStateAddress, StakedStateOpAttributes};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 use parity_scale_codec::{Decode, Encode, Error, Input, Output};
 
 use serde::{Deserialize, Serialize};
@@ -43,14 +43,9 @@ impl Encode for UnjailTx {
     }
 }
 
-impl TransactionId for UnjailTx {}
-
-impl TransactionType for UnjailTx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"unjailtx"
+impl From<UnjailTx> for TaggedTransaction {
+    fn from(tx: UnjailTx) -> TaggedTransaction {
+        TaggedTransaction::UnjailTx(tx)
     }
 }
 
diff --git a/chain-core/src/tx/data/mod.rs b/chain-core/src/tx/data/mod.rs
index 3f751a2a..adbcc633 100644
--- a/chain-core/src/tx/data/mod.rs
+++ b/chain-core/src/tx/data/mod.rs
@@ -20,7 +20,7 @@ use serde::{Deserialize, Serialize};
 use crate::common::H256;
 use crate::init::coin::{sum_coins, Coin, CoinError};
 use crate::tx::data::{attribute::TxAttributes, input::TxoPointer, output::TxOut};
-use crate::tx::{TransactionId, TransactionType};
+use crate::tx::TaggedTransaction;
 
 /// Each input is 34 bytes
 /// Each output is 33 (address) + 8 (amount) + 9 (timelock) = 50 bytes
@@ -30,15 +30,6 @@ use crate::tx::{TransactionId, TransactionType};
 /// So, maximum transaction size (34 * 64) + (50 * 64) + 2688 = 8064
 const MAX_TX_SIZE: usize = 8100; // 8100 bytes
 
-/// Calculates hash of the input data -- if SCALE-serialized TX is passed in, it's equivalent to TxId.
-/// It uses blake3.
-pub fn txid_hash<T: TransactionId>(buf: &T) -> H256 {
-    let mut hasher = blake3::Hasher::new();
-    hasher.update(T::tx_type_tag());
-    hasher.update(&buf.encode());
-    hasher.finalize().into()
-}
-
 /// Key to identify the used TXID hash function, e.g. in ProofOps.
 pub const TXID_HASH_ID: &[u8; 6] = b"blake3";
 
@@ -102,14 +93,9 @@ impl fmt::Display for Tx {
     }
 }
 
-impl TransactionId for Tx {}
-
-impl TransactionType for Tx {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8] {
-        b"transfertx"
+impl From<Tx> for TaggedTransaction {
+    fn from(tx: Tx) -> TaggedTransaction {
+        TaggedTransaction::Transfer(tx)
     }
 }
 
diff --git a/chain-core/src/tx/mod.rs b/chain-core/src/tx/mod.rs
index 86104934..47d2a0c8 100644
--- a/chain-core/src/tx/mod.rs
+++ b/chain-core/src/tx/mod.rs
@@ -20,8 +20,6 @@ use crate::state::account::{
 };
 use crate::state::tendermint::BlockHeight;
 use crate::state::validator::NodeJoinRequestTx;
-#[cfg(not(feature = "legacy-txid"))]
-use crate::tx::data::txid_hash;
 use crate::tx::data::TxId;
 use aead::Payload;
 use data::input::{TxoPointer, TxoSize};
@@ -546,29 +544,56 @@ impl Decode for TxAux {
 /// or make sure you override the default implementation to take the identifier
 /// from the transaction data type.
 #[cfg(not(feature = "legacy-txid"))]
-pub trait TransactionId: Encode + Sized + TransactionType {
+pub trait TransactionId {
     /// retrieves a TX ID (currently blake3(<tx type tag> || scale_codec_bytes(tx)))
+    fn id(&self) -> TxId;
+}
+
+impl<T: Into<TaggedTransaction> + Clone> TransactionId for T {
     fn id(&self) -> TxId {
-        txid_hash(self)
+        let tx: TaggedTransaction = self.clone().into();
+        TaggedTransaction::from(tx).id()
     }
 }
 
 /// 0.5-compatible version: This trait is for transaction data type to define the "txid"
 /// that's used as a message digest in signing.
 #[cfg(feature = "legacy-txid")]
-pub trait TransactionId: Encode + Sized + TransactionType {
+pub trait TransactionId: Encode {
     /// 0.5-compatible version: retrieves a TX ID (currently blake3(scale_codec_bytes(tx)))
     fn id(&self) -> TxId {
         blake3::hash(&self.encode()).into()
     }
 }
 
-/// type tag used in transaction identifier calculations
-pub trait TransactionType {
-    /// NOTE: this tag is used in tx id calculations / consensus critical,
-    /// so even if type is renamed or otherwise refactored,
-    /// this tag MUST remain the same
-    fn tx_type_tag() -> &'static [u8];
+/// used for TXID calculation -- contains all possible tx types
+/// NOTE: do not reorder, as the byte tag is used in txid calculation
+#[derive(Encode)]
+pub enum TaggedTransaction {
+    /// transfer transaction
+    Transfer(Tx),
+    /// deposit stake to bonded amount
+    Deposit(DepositBondTx),
+    /// withdraw unbonded amount
+    Withdraw(WithdrawUnbondedTx),
+    /// unbond stake
+    UnbondStakeTx(UnbondTx),
+    /// unjail request
+    UnjailTx(UnjailTx),
+    /// node join request
+    NodeJoinTx(NodeJoinRequestTx),
+    /// removal proposals + commit
+    MLSRemoveCommitProposal(crate::mls::CommitRemoveTx),
+    /// update proposal + commit
+    MLSSelfUpdateProposal(crate::mls::SelfUpdateProposalTx),
+    /// NACK
+    MLSMsgNack(crate::mls::NackMsgTx),
+}
+
+impl TaggedTransaction {
+    fn id(&self) -> TxId {
+        blake3::hash(&self.encode()).into()
+    }
 }
 
 impl TxAux {
```

Reply:
> tried it -- just a sketch -- two small annoyances with this one are 1) clone required (I can try to rewrite it with `TaggedTx<'a> { Transfer(&'a ...` if that would work), 2) to keep the old behaviour under a feature-flag, there'll need to be feature guards on all transactions (to import and `impl TransactionId`)
> 
> ```
> diff --git a/chain-core/src/mls/mod.rs b/chain-core/src/mls/mod.rs
> index 02230b3c..e3d05550 100644
> --- a/chain-core/src/mls/mod.rs
> +++ b/chain-core/src/mls/mod.rs
> @@ -1,4 +1,4 @@
> -use crate::tx::{data::TxId, TransactionId, TransactionType};
> +use crate::tx::{data::TxId, TaggedTransaction, TransactionId};
>  use parity_scale_codec::{Decode, Encode};
>  use std::prelude::v1::Vec;
>  
> @@ -9,14 +9,9 @@ pub struct CommitRemoveTx {
>      commit: Vec<u8>,         // MLSPlaintext -- Commit
>  }
>  
> -impl TransactionId for CommitRemoveTx {}
> -
> -impl TransactionType for CommitRemoveTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"commitremovetx"
> +impl From<CommitRemoveTx> for TaggedTransaction {
> +    fn from(tx: CommitRemoveTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSRemoveCommitProposal(tx)
>      }
>  }
>  
> @@ -27,14 +22,9 @@ pub struct SelfUpdateProposalTx {
>      commit: Vec<u8>,   // MLSPlaintext -- Commit
>  }
>  
> -impl TransactionId for SelfUpdateProposalTx {}
> -
> -impl TransactionType for SelfUpdateProposalTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same    
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"selfupdateproposaltx"
> +impl From<SelfUpdateProposalTx> for TaggedTransaction {
> +    fn from(tx: SelfUpdateProposalTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSSelfUpdateProposal(tx)
>      }
>  }
>  
> @@ -42,14 +32,9 @@ impl TransactionType for SelfUpdateProposalTx {
>  #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
>  pub struct NackMsgTx(Vec<u8>);
>  
> -impl TransactionId for NackMsgTx {}
> -
> -impl TransactionType for NackMsgTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same    
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"nackmsgtx"
> +impl From<NackMsgTx> for TaggedTransaction {
> +    fn from(tx: NackMsgTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSMsgNack(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/deposit.rs b/chain-core/src/state/account/op/data/deposit.rs
> index 15030393..eae4a6cd 100644
> --- a/chain-core/src/state/account/op/data/deposit.rs
> +++ b/chain-core/src/state/account/op/data/deposit.rs
> @@ -1,7 +1,7 @@
>  use crate::state::account::address::StakedStateAddress;
>  use crate::state::account::op::data::attribute::StakedStateOpAttributes;
>  use crate::tx::data::input::TxoPointer;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  use serde::{Deserialize, Serialize};
>  use std::fmt;
> @@ -59,14 +59,9 @@ impl Encode for DepositBondTx {
>      }
>  }
>  
> -impl TransactionId for DepositBondTx {}
> -
> -impl TransactionType for DepositBondTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"deposittx"
> +impl From<DepositBondTx> for TaggedTransaction {
> +    fn from(tx: DepositBondTx) -> TaggedTransaction {
> +        TaggedTransaction::Deposit(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/unbond.rs b/chain-core/src/state/account/op/data/unbond.rs
> index 14606dd0..2043af6b 100644
> --- a/chain-core/src/state/account/op/data/unbond.rs
> +++ b/chain-core/src/state/account/op/data/unbond.rs
> @@ -2,7 +2,7 @@ use crate::init::coin::Coin;
>  use crate::state::account::address::StakedStateAddress;
>  use crate::state::account::op::data::attribute::StakedStateOpAttributes;
>  use crate::state::account::Nonce;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -55,14 +55,9 @@ impl Encode for UnbondTx {
>      }
>  }
>  
> -impl TransactionId for UnbondTx {}
> -
> -impl TransactionType for UnbondTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"unbondtx"
> +impl From<UnbondTx> for TaggedTransaction {
> +    fn from(tx: UnbondTx) -> TaggedTransaction {
> +        TaggedTransaction::UnbondStakeTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/withdraw.rs b/chain-core/src/state/account/op/data/withdraw.rs
> index d3e0c3d8..43bb84a4 100644
> --- a/chain-core/src/state/account/op/data/withdraw.rs
> +++ b/chain-core/src/state/account/op/data/withdraw.rs
> @@ -2,7 +2,7 @@ use crate::init::coin::{sum_coins, Coin, CoinError};
>  use crate::state::account::Nonce;
>  use crate::tx::data::attribute::TxAttributes;
>  use crate::tx::data::output::TxOut;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -48,14 +48,9 @@ impl Encode for WithdrawUnbondedTx {
>      }
>  }
>  
> -impl TransactionId for WithdrawUnbondedTx {}
> -
> -impl TransactionType for WithdrawUnbondedTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"withdrawtx"
> +impl From<WithdrawUnbondedTx> for TaggedTransaction {
> +    fn from(tx: WithdrawUnbondedTx) -> TaggedTransaction {
> +        TaggedTransaction::Withdraw(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/validator/nodejoin.rs b/chain-core/src/state/validator/nodejoin.rs
> index 3f690dfd..514f144c 100644
> --- a/chain-core/src/state/validator/nodejoin.rs
> +++ b/chain-core/src/state/validator/nodejoin.rs
> @@ -1,5 +1,5 @@
>  use crate::state::account::{NodeMetadata, Nonce, StakedStateAddress, StakedStateOpAttributes};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -60,14 +60,9 @@ impl Encode for NodeJoinRequestTx {
>      }
>  }
>  
> -impl TransactionId for NodeJoinRequestTx {}
> -
> -impl TransactionType for NodeJoinRequestTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"nodejointx"
> +impl From<NodeJoinRequestTx> for TaggedTransaction {
> +    fn from(tx: NodeJoinRequestTx) -> TaggedTransaction {
> +        TaggedTransaction::NodeJoinTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/validator/unjail.rs b/chain-core/src/state/validator/unjail.rs
> index 6e38d00c..f3b17d47 100644
> --- a/chain-core/src/state/validator/unjail.rs
> +++ b/chain-core/src/state/validator/unjail.rs
> @@ -1,5 +1,5 @@
>  use crate::state::account::{Nonce, StakedStateAddress, StakedStateOpAttributes};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -43,14 +43,9 @@ impl Encode for UnjailTx {
>      }
>  }
>  
> -impl TransactionId for UnjailTx {}
> -
> -impl TransactionType for UnjailTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"unjailtx"
> +impl From<UnjailTx> for TaggedTransaction {
> +    fn from(tx: UnjailTx) -> TaggedTransaction {
> +        TaggedTransaction::UnjailTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/tx/data/mod.rs b/chain-core/src/tx/data/mod.rs
> index 3f751a2a..adbcc633 100644
> --- a/chain-core/src/tx/data/mod.rs
> +++ b/chain-core/src/tx/data/mod.rs
> @@ -20,7 +20,7 @@ use serde::{Deserialize, Serialize};
>  use crate::common::H256;
>  use crate::init::coin::{sum_coins, Coin, CoinError};
>  use crate::tx::data::{attribute::TxAttributes, input::TxoPointer, output::TxOut};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  
>  /// Each input is 34 bytes
>  /// Each output is 33 (address) + 8 (amount) + 9 (timelock) = 50 bytes
> @@ -30,15 +30,6 @@ use crate::tx::{TransactionId, TransactionType};
>  /// So, maximum transaction size (34 * 64) + (50 * 64) + 2688 = 8064
>  const MAX_TX_SIZE: usize = 8100; // 8100 bytes
>  
> -/// Calculates hash of the input data -- if SCALE-serialized TX is passed in, it's equivalent to TxId.
> -/// It uses blake3.
> -pub fn txid_hash<T: TransactionId>(buf: &T) -> H256 {
> -    let mut hasher = blake3::Hasher::new();
> -    hasher.update(T::tx_type_tag());
> -    hasher.update(&buf.encode());
> -    hasher.finalize().into()
> -}
> -
>  /// Key to identify the used TXID hash function, e.g. in ProofOps.
>  pub const TXID_HASH_ID: &[u8; 6] = b"blake3";
>  
> @@ -102,14 +93,9 @@ impl fmt::Display for Tx {
>      }
>  }
>  
> -impl TransactionId for Tx {}
> -
> -impl TransactionType for Tx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"transfertx"
> +impl From<Tx> for TaggedTransaction {
> +    fn from(tx: Tx) -> TaggedTransaction {
> +        TaggedTransaction::Transfer(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/tx/mod.rs b/chain-core/src/tx/mod.rs
> index 86104934..47d2a0c8 100644
> --- a/chain-core/src/tx/mod.rs
> +++ b/chain-core/src/tx/mod.rs
> @@ -20,8 +20,6 @@ use crate::state::account::{
>  };
>  use crate::state::tendermint::BlockHeight;
>  use crate::state::validator::NodeJoinRequestTx;
> -#[cfg(not(feature = "legacy-txid"))]
> -use crate::tx::data::txid_hash;
>  use crate::tx::data::TxId;
>  use aead::Payload;
>  use data::input::{TxoPointer, TxoSize};
> @@ -546,29 +544,56 @@ impl Decode for TxAux {
>  /// or make sure you override the default implementation to take the identifier
>  /// from the transaction data type.
>  #[cfg(not(feature = "legacy-txid"))]
> -pub trait TransactionId: Encode + Sized + TransactionType {
> +pub trait TransactionId {
>      /// retrieves a TX ID (currently blake3(<tx type tag> || scale_codec_bytes(tx)))
> +    fn id(&self) -> TxId;
> +}
> +
> +impl<T: Into<TaggedTransaction> + Clone> TransactionId for T {
>      fn id(&self) -> TxId {
> -        txid_hash(self)
> +        let tx: TaggedTransaction = self.clone().into();
> +        TaggedTransaction::from(tx).id()
>      }
>  }
>  
>  /// 0.5-compatible version: This trait is for transaction data type to define the "txid"
>  /// that's used as a message digest in signing.
>  #[cfg(feature = "legacy-txid")]
> -pub trait TransactionId: Encode + Sized + TransactionType {
> +pub trait TransactionId: Encode {
>      /// 0.5-compatible version: retrieves a TX ID (currently blake3(scale_codec_bytes(tx)))
>      fn id(&self) -> TxId {
>          blake3::hash(&self.encode()).into()
>      }
>  }
>  
> -/// type tag used in transaction identifier calculations
> -pub trait TransactionType {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8];
> +/// used for TXID calculation -- contains all possible tx types
> +/// NOTE: do not reorder, as the byte tag is used in txid calculation
> +#[derive(Encode)]
> +pub enum TaggedTransaction {
> +    /// transfer transaction
> +    Transfer(Tx),
> +    /// deposit stake to bonded amount
> +    Deposit(DepositBondTx),
> +    /// withdraw unbonded amount
> +    Withdraw(WithdrawUnbondedTx),
> +    /// unbond stake
> +    UnbondStakeTx(UnbondTx),
> +    /// unjail request
> +    UnjailTx(UnjailTx),
> +    /// node join request
> +    NodeJoinTx(NodeJoinRequestTx),
> +    /// removal proposals + commit
> +    MLSRemoveCommitProposal(crate::mls::CommitRemoveTx),
> +    /// update proposal + commit
> +    MLSSelfUpdateProposal(crate::mls::SelfUpdateProposalTx),
> +    /// NACK
> +    MLSMsgNack(crate::mls::NackMsgTx),
> +}
> +
> +impl TaggedTransaction {
> +    fn id(&self) -> TxId {
> +        blake3::hash(&self.encode()).into()
> +    }
>  }
>  
>  impl TxAux {
> tomas@MacBook-Pro chain-core % git diff > 
> tomas@MacBook-Pro chain-core % cat test-single.patch 
> diff --git a/chain-core/src/mls/mod.rs b/chain-core/src/mls/mod.rs
> index 02230b3c..e3d05550 100644
> --- a/chain-core/src/mls/mod.rs
> +++ b/chain-core/src/mls/mod.rs
> @@ -1,4 +1,4 @@
> -use crate::tx::{data::TxId, TransactionId, TransactionType};
> +use crate::tx::{data::TxId, TaggedTransaction, TransactionId};
>  use parity_scale_codec::{Decode, Encode};
>  use std::prelude::v1::Vec;
>  
> @@ -9,14 +9,9 @@ pub struct CommitRemoveTx {
>      commit: Vec<u8>,         // MLSPlaintext -- Commit
>  }
>  
> -impl TransactionId for CommitRemoveTx {}
> -
> -impl TransactionType for CommitRemoveTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"commitremovetx"
> +impl From<CommitRemoveTx> for TaggedTransaction {
> +    fn from(tx: CommitRemoveTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSRemoveCommitProposal(tx)
>      }
>  }
>  
> @@ -27,14 +22,9 @@ pub struct SelfUpdateProposalTx {
>      commit: Vec<u8>,   // MLSPlaintext -- Commit
>  }
>  
> -impl TransactionId for SelfUpdateProposalTx {}
> -
> -impl TransactionType for SelfUpdateProposalTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same    
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"selfupdateproposaltx"
> +impl From<SelfUpdateProposalTx> for TaggedTransaction {
> +    fn from(tx: SelfUpdateProposalTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSSelfUpdateProposal(tx)
>      }
>  }
>  
> @@ -42,14 +32,9 @@ impl TransactionType for SelfUpdateProposalTx {
>  #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
>  pub struct NackMsgTx(Vec<u8>);
>  
> -impl TransactionId for NackMsgTx {}
> -
> -impl TransactionType for NackMsgTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same    
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"nackmsgtx"
> +impl From<NackMsgTx> for TaggedTransaction {
> +    fn from(tx: NackMsgTx) -> TaggedTransaction {
> +        TaggedTransaction::MLSMsgNack(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/deposit.rs b/chain-core/src/state/account/op/data/deposit.rs
> index 15030393..eae4a6cd 100644
> --- a/chain-core/src/state/account/op/data/deposit.rs
> +++ b/chain-core/src/state/account/op/data/deposit.rs
> @@ -1,7 +1,7 @@
>  use crate::state::account::address::StakedStateAddress;
>  use crate::state::account::op::data::attribute::StakedStateOpAttributes;
>  use crate::tx::data::input::TxoPointer;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  use serde::{Deserialize, Serialize};
>  use std::fmt;
> @@ -59,14 +59,9 @@ impl Encode for DepositBondTx {
>      }
>  }
>  
> -impl TransactionId for DepositBondTx {}
> -
> -impl TransactionType for DepositBondTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"deposittx"
> +impl From<DepositBondTx> for TaggedTransaction {
> +    fn from(tx: DepositBondTx) -> TaggedTransaction {
> +        TaggedTransaction::Deposit(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/unbond.rs b/chain-core/src/state/account/op/data/unbond.rs
> index 14606dd0..2043af6b 100644
> --- a/chain-core/src/state/account/op/data/unbond.rs
> +++ b/chain-core/src/state/account/op/data/unbond.rs
> @@ -2,7 +2,7 @@ use crate::init::coin::Coin;
>  use crate::state::account::address::StakedStateAddress;
>  use crate::state::account::op::data::attribute::StakedStateOpAttributes;
>  use crate::state::account::Nonce;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -55,14 +55,9 @@ impl Encode for UnbondTx {
>      }
>  }
>  
> -impl TransactionId for UnbondTx {}
> -
> -impl TransactionType for UnbondTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"unbondtx"
> +impl From<UnbondTx> for TaggedTransaction {
> +    fn from(tx: UnbondTx) -> TaggedTransaction {
> +        TaggedTransaction::UnbondStakeTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/account/op/data/withdraw.rs b/chain-core/src/state/account/op/data/withdraw.rs
> index d3e0c3d8..43bb84a4 100644
> --- a/chain-core/src/state/account/op/data/withdraw.rs
> +++ b/chain-core/src/state/account/op/data/withdraw.rs
> @@ -2,7 +2,7 @@ use crate::init::coin::{sum_coins, Coin, CoinError};
>  use crate::state::account::Nonce;
>  use crate::tx::data::attribute::TxAttributes;
>  use crate::tx::data::output::TxOut;
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -48,14 +48,9 @@ impl Encode for WithdrawUnbondedTx {
>      }
>  }
>  
> -impl TransactionId for WithdrawUnbondedTx {}
> -
> -impl TransactionType for WithdrawUnbondedTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"withdrawtx"
> +impl From<WithdrawUnbondedTx> for TaggedTransaction {
> +    fn from(tx: WithdrawUnbondedTx) -> TaggedTransaction {
> +        TaggedTransaction::Withdraw(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/validator/nodejoin.rs b/chain-core/src/state/validator/nodejoin.rs
> index 3f690dfd..514f144c 100644
> --- a/chain-core/src/state/validator/nodejoin.rs
> +++ b/chain-core/src/state/validator/nodejoin.rs
> @@ -1,5 +1,5 @@
>  use crate::state::account::{NodeMetadata, Nonce, StakedStateAddress, StakedStateOpAttributes};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -60,14 +60,9 @@ impl Encode for NodeJoinRequestTx {
>      }
>  }
>  
> -impl TransactionId for NodeJoinRequestTx {}
> -
> -impl TransactionType for NodeJoinRequestTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"nodejointx"
> +impl From<NodeJoinRequestTx> for TaggedTransaction {
> +    fn from(tx: NodeJoinRequestTx) -> TaggedTransaction {
> +        TaggedTransaction::NodeJoinTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/state/validator/unjail.rs b/chain-core/src/state/validator/unjail.rs
> index 6e38d00c..f3b17d47 100644
> --- a/chain-core/src/state/validator/unjail.rs
> +++ b/chain-core/src/state/validator/unjail.rs
> @@ -1,5 +1,5 @@
>  use crate::state::account::{Nonce, StakedStateAddress, StakedStateOpAttributes};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  use parity_scale_codec::{Decode, Encode, Error, Input, Output};
>  
>  use serde::{Deserialize, Serialize};
> @@ -43,14 +43,9 @@ impl Encode for UnjailTx {
>      }
>  }
>  
> -impl TransactionId for UnjailTx {}
> -
> -impl TransactionType for UnjailTx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"unjailtx"
> +impl From<UnjailTx> for TaggedTransaction {
> +    fn from(tx: UnjailTx) -> TaggedTransaction {
> +        TaggedTransaction::UnjailTx(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/tx/data/mod.rs b/chain-core/src/tx/data/mod.rs
> index 3f751a2a..adbcc633 100644
> --- a/chain-core/src/tx/data/mod.rs
> +++ b/chain-core/src/tx/data/mod.rs
> @@ -20,7 +20,7 @@ use serde::{Deserialize, Serialize};
>  use crate::common::H256;
>  use crate::init::coin::{sum_coins, Coin, CoinError};
>  use crate::tx::data::{attribute::TxAttributes, input::TxoPointer, output::TxOut};
> -use crate::tx::{TransactionId, TransactionType};
> +use crate::tx::TaggedTransaction;
>  
>  /// Each input is 34 bytes
>  /// Each output is 33 (address) + 8 (amount) + 9 (timelock) = 50 bytes
> @@ -30,15 +30,6 @@ use crate::tx::{TransactionId, TransactionType};
>  /// So, maximum transaction size (34 * 64) + (50 * 64) + 2688 = 8064
>  const MAX_TX_SIZE: usize = 8100; // 8100 bytes
>  
> -/// Calculates hash of the input data -- if SCALE-serialized TX is passed in, it's equivalent to TxId.
> -/// It uses blake3.
> -pub fn txid_hash<T: TransactionId>(buf: &T) -> H256 {
> -    let mut hasher = blake3::Hasher::new();
> -    hasher.update(T::tx_type_tag());
> -    hasher.update(&buf.encode());
> -    hasher.finalize().into()
> -}
> -
>  /// Key to identify the used TXID hash function, e.g. in ProofOps.
>  pub const TXID_HASH_ID: &[u8; 6] = b"blake3";
>  
> @@ -102,14 +93,9 @@ impl fmt::Display for Tx {
>      }
>  }
>  
> -impl TransactionId for Tx {}
> -
> -impl TransactionType for Tx {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8] {
> -        b"transfertx"
> +impl From<Tx> for TaggedTransaction {
> +    fn from(tx: Tx) -> TaggedTransaction {
> +        TaggedTransaction::Transfer(tx)
>      }
>  }
>  
> diff --git a/chain-core/src/tx/mod.rs b/chain-core/src/tx/mod.rs
> index 86104934..47d2a0c8 100644
> --- a/chain-core/src/tx/mod.rs
> +++ b/chain-core/src/tx/mod.rs
> @@ -20,8 +20,6 @@ use crate::state::account::{
>  };
>  use crate::state::tendermint::BlockHeight;
>  use crate::state::validator::NodeJoinRequestTx;
> -#[cfg(not(feature = "legacy-txid"))]
> -use crate::tx::data::txid_hash;
>  use crate::tx::data::TxId;
>  use aead::Payload;
>  use data::input::{TxoPointer, TxoSize};
> @@ -546,29 +544,56 @@ impl Decode for TxAux {
>  /// or make sure you override the default implementation to take the identifier
>  /// from the transaction data type.
>  #[cfg(not(feature = "legacy-txid"))]
> -pub trait TransactionId: Encode + Sized + TransactionType {
> +pub trait TransactionId {
>      /// retrieves a TX ID (currently blake3(<tx type tag> || scale_codec_bytes(tx)))
> +    fn id(&self) -> TxId;
> +}
> +
> +impl<T: Into<TaggedTransaction> + Clone> TransactionId for T {
>      fn id(&self) -> TxId {
> -        txid_hash(self)
> +        let tx: TaggedTransaction = self.clone().into();
> +        TaggedTransaction::from(tx).id()
>      }
>  }
>  
>  /// 0.5-compatible version: This trait is for transaction data type to define the "txid"
>  /// that's used as a message digest in signing.
>  #[cfg(feature = "legacy-txid")]
> -pub trait TransactionId: Encode + Sized + TransactionType {
> +pub trait TransactionId: Encode {
>      /// 0.5-compatible version: retrieves a TX ID (currently blake3(scale_codec_bytes(tx)))
>      fn id(&self) -> TxId {
>          blake3::hash(&self.encode()).into()
>      }
>  }
>  
> -/// type tag used in transaction identifier calculations
> -pub trait TransactionType {
> -    /// NOTE: this tag is used in tx id calculations / consensus critical,
> -    /// so even if type is renamed or otherwise refactored,
> -    /// this tag MUST remain the same
> -    fn tx_type_tag() -> &'static [u8];
> +/// used for TXID calculation -- contains all possible tx types
> +/// NOTE: do not reorder, as the byte tag is used in txid calculation
> +#[derive(Encode)]
> +pub enum TaggedTransaction {
> +    /// transfer transaction
> +    Transfer(Tx),
> +    /// deposit stake to bonded amount
> +    Deposit(DepositBondTx),
> +    /// withdraw unbonded amount
> +    Withdraw(WithdrawUnbondedTx),
> +    /// unbond stake
> +    UnbondStakeTx(UnbondTx),
> +    /// unjail request
> +    UnjailTx(UnjailTx),
> +    /// node join request
> +    NodeJoinTx(NodeJoinRequestTx),
> +    /// removal proposals + commit
> +    MLSRemoveCommitProposal(crate::mls::CommitRemoveTx),
> +    /// update proposal + commit
> +    MLSSelfUpdateProposal(crate::mls::SelfUpdateProposalTx),
> +    /// NACK
> +    MLSMsgNack(crate::mls::NackMsgTx),
> +}
> +
> +impl TaggedTransaction {
> +    fn id(&self) -> TxId {
> +        blake3::hash(&self.encode()).into()
> +    }
>  }
>  
>  impl TxAux {
> ```

do we have to keep the old behaviour?

Reply:
@yihuang yes, at least for a month, as there's already external code @linfeng-crypto needs to integrate with and it calculates TXID as before